compiler/env/newmemory/DebugAllocator.cpp

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2017, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX)
+#include <sys/mman.h>
+#if defined(__APPLE__) || !defined(MAP_ANONYMOUS)
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+#elif defined(OMR_OS_WINDOWS)
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#else
+#include <string.h>
+#endif /* defined(OMR_OS_WINDOWS) */
+
+#include <stddef.h>  // for size_t
+#include <cstdlib>   // for free, malloc
+#include <new>       // for bad_alloc, nothrow, nothrow_t
+#include "infra/Assert.hpp"
+#include "env/mem/DebugAllocator.hpp"
+
+void *
+TR::DebugAllocator::allocate(size_t size, const std::nothrow_t tag,void * hint) const throw()
+   {
+#if (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX)
+   void *p = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+   return (p == MAP_FAILED) ? NULL : p;
+#elif defined(OMR_OS_WINDOWS)
+   return  VirtualAlloc(NULL, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+#else
+   return this->TR::RawAllocator::allocate(size, hint);
+#endif /* (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX) */
+   }
+
+void
+TR::DebugAllocator::deallocate(void* p) const throw()
+   {
+#if (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX)
+   TR_ASSERT(0, "DebugAllocator does not support deallocation without size parameter on this platform");
```

Comment:
It seems like this could be problematic when debugging. Would it be possible to make it so that users must always pass a size and avoid this problem altogether?

Reply:
it's the way the code works right now. We can change how it works later; I would prefer to keep it as functionally similar as possible for now.